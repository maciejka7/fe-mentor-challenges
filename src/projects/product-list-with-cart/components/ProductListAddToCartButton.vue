<script setup lang="ts">
import { CircleMinus, CirclePlus, ShoppingCart } from 'lucide-vue-next'
import { defineEmits, defineProps } from 'vue'

interface ProductListButtonProps {
  label: string
}

type ProductListButtonState = 'active' | 'idle'

const { label } = defineProps<ProductListButtonProps>()
const emit = defineEmits<(e: 'addAmount', amount: number) => void>()
const buttonState = ref<ProductListButtonState>('idle')
const quantity = ref<number>(0)

function toggleButtonState() {
  if (buttonState.value === 'active') {
    buttonState.value = 'idle'
  }
  else {
    buttonState.value = 'active'
  }
}

function changeQuantity(value: number) {
  if (!(quantity.value === 0 && value === -1)) {
    quantity.value += value
    emit('addAmount', quantity.value)
  }
}
</script>

<template>
  <button
    v-if="buttonState === 'idle'" data-test="idle-button"
    class="border-plwc-rose-400 hover:border-plwc-red hover:text-plwc-red relative min-w-40 w-fit flex gap-2 border rounded-full bg-white px-7 py-3 transition-all transition-duration-300"
    @click="toggleButtonState"
  >
    <ShoppingCart class="text-plwc-red" /> {{ label }}
  </button>
  <button
    v-else data-test="active-button"
    class="border-plwc-rose-400 hover:border-plwc-red bg-plwc-red relative min-w-40 w-fit flex justify-between gap-2 border rounded-full p-3 text-white transition-all transition-duration-300"
    @click="toggleButtonState"
  >
    <CircleMinus data-test="active-button-subtract" @click.stop="changeQuantity(-1)" />
    {{ quantity }}
    <CirclePlus data-test="active-button-add" @click.stop="changeQuantity(1)" />
  </button>
</template>

<style scoped></style>
