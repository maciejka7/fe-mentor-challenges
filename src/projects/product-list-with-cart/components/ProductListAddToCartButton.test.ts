import type { VueWrapper } from '@vue/test-utils'
import { mount } from '@vue/test-utils'
import { describe, expect, it } from 'vitest'
import ProductListAddToCartButton from './ProductListAddToCartButton.vue'

const buttonComponentsElements = {
  idleButton(wrapper: VueWrapper) { return wrapper.get('[data-test="idle-button"]') },
  activeButton(wrapper: VueWrapper) { return wrapper.get('[data-test="active-button"]') },
  activeButtonAdd(wrapper: VueWrapper) { return wrapper.get('[data-test="active-button-add"]') },
  activeButtonSubtract(wrapper: VueWrapper) { return wrapper.get('[data-test="active-button-subtract"]') },
}

describe('component ProductListButton.vue', () => {
  it('should render', () => {
    const wrapper = mount(ProductListAddToCartButton, { props: { label: 'Add to cart' } })
    expect(wrapper.text()).toContain('Add to cart')
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should change state after click', async () => {
    const wrapper = mount(ProductListAddToCartButton, { props: { label: 'Add to cart' } })
    const idleButton = buttonComponentsElements.idleButton(wrapper)
    await idleButton.trigger('click')
    expect(wrapper.text()).toContain('0')
    await idleButton.trigger('click')
    expect(wrapper.text()).toContain('Add to cart')
  })
  it('should change count after plus button click', async () => {
    const wrapper = mount(ProductListAddToCartButton, { props: { label: 'Add to cart' } })
    const idleButton = buttonComponentsElements.idleButton(wrapper)
    await idleButton.trigger('click')

    const plusButton = buttonComponentsElements.activeButtonAdd(wrapper)

    await plusButton.trigger('click')
    await plusButton.trigger('click')

    expect(wrapper.text()).toContain('2')
  })
  it('should change count after minus button click', async () => {
    const wrapper = mount(ProductListAddToCartButton, { props: { label: 'Add to cart' } })
    const idleButton = buttonComponentsElements.idleButton(wrapper)
    await idleButton.trigger('click')

    const plusButton = buttonComponentsElements.activeButtonAdd(wrapper)
    const minusButton = buttonComponentsElements.activeButtonSubtract(wrapper)

    await plusButton.trigger('click')
    await plusButton.trigger('click')
    await minusButton.trigger('click')

    expect(wrapper.text()).toContain('1')
  })
  it('should count be at least 0 after minus clicking', async () => {
    const wrapper = mount(ProductListAddToCartButton, { props: { label: 'Add to cart' } })
    const idleButton = buttonComponentsElements.idleButton(wrapper)
    await idleButton.trigger('click')

    const plusButton = buttonComponentsElements.activeButtonAdd(wrapper)
    const minusButton = buttonComponentsElements.activeButtonSubtract(wrapper)

    await plusButton.trigger('click')
    await minusButton.trigger('click')
    await minusButton.trigger('click')
    await minusButton.trigger('click')
    await minusButton.trigger('click')

    expect(wrapper.text()).toContain('0')
  })
})
