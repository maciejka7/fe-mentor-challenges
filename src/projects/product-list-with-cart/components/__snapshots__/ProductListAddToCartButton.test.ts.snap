// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`component ProductListButton.vue > should render 1`] = `
"<button data-test="idle-button" class="border-plwc-rose-400 w-fit relative hover:border-plwc-red hover:text-plwc-red min-w-40 flex gap-2 border rounded-full bg-white px-7 py-3 transition-all transition-duration-300"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-shopping-cart-icon text-plwc-red">
    <circle cx="8" cy="21" r="1"></circle>
    <circle cx="19" cy="21" r="1"></circle>
    <path d="M2.05 2.05h2l2.66 12.42a2 2 0 0 0 2 1.58h9.78a2 2 0 0 0 1.95-1.57l1.65-7.43H5.12"></path>
  </svg> Add to cart</button>"
`;
